File: src/main/kotlin/io/energyconsumptionoptimizer/mapservice/domain/ports/ZoneService.kt
package io.energyconsumptionoptimizer.mapservice.domain.ports

import io.energyconsumptionoptimizer.mapservice.domain.Zone
import io.energyconsumptionoptimizer.mapservice.domain.ZoneID

/**
 * Service interface for managing zone operations.
 */
interface ZoneService {
    /**
     * Creates a new zone with the specified properties.
     *
     * @param name The human-readable name of the zone
     * @param colorHex The hexadecimal color code for visual representation
     * @param vertices The list of (x, y) coordinate pairs defining the zone's vertices
     * @return The created zone entity
     *
     * @throws io.energyconsumptionoptimizer.mapservice.domain.errors.ZoneNameAlreadyExistsException
     * @throws IllegalArgumentException if the vertices do not form a valid polygon, if the name is blank or if the color format is invalid
     */
    suspend fun createZone(
        name: String,
        colorHex: String,
        vertices: List<Pair<Double, Double>>,
    ): Zone

    /**
     * Retrieves all zones defined within the floor plan.
     *
     * @return A list of all zones, or an empty list if no zones exist
     */
    suspend fun getZones(): List<Zone>

    /**
     * Retrieves a specific zone by its unique identifier.
     *
     * @param id The unique identifier of the zone
     * @return The zone if found, null otherwise
     */
    suspend fun getZone(id: ZoneID): Zone?

    /**
     * Updates an existing zone's properties.
     *
     * Only non-null parameters will be updated. If a parameter is null, the existing value is retained.
     *
     * @param id The unique identifier of the zone to update
     * @param name The new name for the zone, or null to keep the existing name
     * @param colorHex The new hexadecimal color code, or null to keep the existing color
     * @param vertices The new list of coordinate pairs defining the boundary, or null to keep the existing boundary
     * @return The updated zone entity
     * @throws io.energyconsumptionoptimizer.mapservice.domain.errors.ZoneIDNotFoundException
     * @throws IllegalArgumentException if the new vertices do not form a valid polygon or if the color format is invalid
     */
    suspend fun updateZone(
        id: ZoneID,
        name: String? = null,
        colorHex: String? = null,
        vertices: List<Pair<Double, Double>>? = null,
    ): Zone

    /**
     * Deletes a zone from the system.
     *
     * Note: This operation should affect smart furniture hookups assigned to this zone.
     *
     * @param id The unique identifier of the zone to delete
     * @throws NoSuchElementException if the zone does not exist
     */
    suspend fun deleteZone(id: ZoneID)
}

File: src/application/inbound/ZoneService.ts
import type { Zone } from "@domain/entities/Zone";

/**
 * @inbound
 * Service interface for managing zone operations.
 */
export interface ZoneService {
  /**
   * Creates a new zone with the given name, color, and polygon boundary.
   *
   * After creation, any existing smart furniture hookups whose position falls
   * inside the new zone are automatically assigned to it.
   *
   * @returns The saved Zone, or an Error for validation failures
   *          (ZoneNameEmptyError, InvalidColorError, InvalidPolygonError,
   *          ZoneNameAlreadyExistsError).
   */
  createZone(
    name: string,
    colorHex: string,
    vertices: [number, number][],
  ): Promise<Zone | Error>;

  /**
   * Retrieves all zones defined within the floor plan.
   */
  getZones(): Promise<Zone[]>;

  /**
   * Retrieves a specific zone by its unique identifier.
   *
   * @returns The Zone if found, `null` otherwise.
   */
  getZone(id: string): Promise<Zone | null>;

  /**
   * Updates an existing zone's name, color, and/or boundary.
   *
   * Only supplied (non-undefined) parameters are updated.
   * When vertices are changed, smart furniture hookup zone assignments are
   * recalculated automatically.
   *
   * @returns The updated Zone, or an Error (ZoneNotFoundError, validation errors).
   */
  updateZone(
    id: string,
    name?: string,
    colorHex?: string,
    vertices?: [number, number][],
  ): Promise<Zone | Error>;

  /**
   * Deletes a zone and unassigns any smart furniture hookups that were inside it.
   * Also removes the zone's tag from the monitoring service measurements.
   *
   * @returns `undefined` on success, or an Error (ZoneNotFoundError, IdEmptyError).
   */
  deleteZone(id: string): Promise<undefined | Error>;
}
